sgkit/stats/regenie.py

```diff
@@ -0,0 +1,778 @@
+        # Extract outcome vector
+        YB = Y[:, [i]]
+        assert XPB.ndim == YB.ndim == 2
+        # Fit and predict folds for each parameter
+        BB, YPB = _ridge_regression_cv(XPB, YB, alphas, n_zero_reg=n_covar)[-2:]
+        assert_array_shape(BB, n_alpha_2, n_sample_block * n_indvar, 1)
+        assert_array_shape(YPB, n_alpha_2, n_sample, 1)
+        BR.append(BB)
+        YR.append(YPB)
+
+    # Concatenate predictions along outcome dimension
+    YR = da.concatenate(YR, axis=2)
+    assert_block_shape(YR, 1, n_sample_block, n_outcome)
+    assert_chunk_shape(YR, n_alpha_2, sample_chunks[0], 1)
+    assert_array_shape(YR, n_alpha_2, n_sample, n_outcome)
+    # Move samples to last dim so all others are batch
+    # dims for R2 calculations
+    YR = da.transpose(YR, (0, 2, 1))
+    assert_array_shape(YR, n_alpha_2, n_outcome, n_sample)
+    YR = YR.rechunk((-1, -1, None))
+    assert_block_shape(YR, 1, 1, n_sample_block)
+    assert YR.shape[1:] == Y.T.shape
+
+    # Concatenate betas along outcome dimension
+    BR = da.concatenate(BR, axis=2)
+    assert_block_shape(BR, 1, n_sample_block, n_outcome)
+    assert_chunk_shape(BR, n_alpha_2, n_indvar, 1)
+    assert_array_shape(BR, n_alpha_2, n_sample_block * n_indvar, n_outcome)
+
+    # Compute R2 scores within each sample block for each outcome + alpha
+    R2 = da.stack(
+        [
+            r2_score(YR.blocks[..., i], Y.T.blocks[..., i])
+            # Avoid warnings on R2 calculations for blocks with single rows
+            if YR.chunks[-1][i] > 1 else da.full(YR.shape[:-1], np.nan)
+            for i in range(n_sample_block)
+        ]
+    )
+    assert_array_shape(R2, n_sample_block, n_alpha_2, n_outcome)
+    # Coerce to finite or nan before nan-aware mean
+    R2 = da.where(da.isfinite(R2), R2, np.nan)
+    # Find highest mean alpha score for each outcome across blocks
+    R2M = da.nanmean(R2, axis=0)
+    assert_array_shape(R2M, n_alpha_2, n_outcome)
+    # Identify index for the alpha value with the highest mean score
+    R2I = da.argmax(R2M, axis=0)
+    assert_array_shape(R2I, n_outcome)
+
+    # Choose the predictions corresponding to the model with best score
+    YRM = da.stack([YR[R2I[i], i, :] for i in range(n_outcome)], axis=-1)
+    YRM = YRM.rechunk((None, -1))
+    assert_block_shape(YRM, n_sample_block, 1)
+    assert_chunk_shape(YRM, sample_chunks[0], n_outcome)
+    assert_array_shape(YRM, n_sample, n_outcome)
+    # Choose the betas corresponding to the model with the best score
+    BRM = da.stack([BR[R2I[i], :, i] for i in range(n_outcome)], axis=-1)
+    BRM = BRM.rechunk((None, -1))
+    assert_block_shape(BRM, n_sample_block, 1)
+    assert_chunk_shape(BRM, n_indvar, n_outcome)
+    assert_array_shape(BRM, n_sample_block * n_indvar, n_outcome)
+    return BRM, YRM
+
+
+def _stage_3(
+    B: Array,
+    YP: Array,
+    X: Array,
+    Y: Array,
+    contigs: Array,
+    variant_chunk_start: ndarray,
```

Comment:
Why are certain types `ndarray` vs `ArrayLike` vs `Array`? Should dask's `Array` by added to `ArrayLike` and should we consistently use `ArrayLike` (even if in this case we are forcing Dask).

Reply:
We definitely shouldn't try to use `ArrayLike` for everything because it's difficult (and impossible in some cases) to make a lot of functions work with either kind of array input.  I think it's best to use the least general type that indicates the level of support you're willing to offer in the function.

I do think all of our top-level functions should use ArrayLike though and that they should be responsible for coercing from there if necessary.  Internal functions need some more freedom though.